pycqed/measurement/waveform_control/pulsar.py

```diff
@@ -1291,33 +1638,199 @@ def program_awgs(self, sequence, awgs='all'):
+            log.debug(f'Start of waveform generation sequence {sequence.name} '
+                     f'{time.time() - t0}')
+            waveforms, _ = sequence.generate_waveforms_sequences(
+                awgs_to_program + awgs_with_channels_to_upload,
+                resolve_segments=False)
+            log.debug(f'End of waveform generation sequence {sequence.name} '
+                     f'{time.time() - t0}')
+            # Check for which channels the sequence structure, or some element
+            # length has changed.
+            # If placeholder waveforms are used, only those channels (and
+            # channels in the same group) will be re-programmed, while other
+            # channels can be re-uploaded by replacing the existing waveforms.
+            ch_length = {}
+            for ch, hashes in channel_hashes.items():
+                ch_awg = self.get(f'{ch}_awg')
+                if ch_awg in awgs_to_program + awgs_with_channels_to_upload:
+                    ch_length[ch] = {
+                        elname: {cw: len(waveforms[h]) for cw, h in el.items()}
+                        for elname, el in hashes.items()}
+                # Checking whether programming is done only for channels that
+                # are marked to be uploaded but not yet marked to be programmed
+                if ch not in channels_to_upload or ch in channels_to_program \
+                        or ch_awg in awgs_to_program:
+                    continue
+                try:
+                    np.testing.assert_equal(
+                        sequence_cache['length'].get(ch, {}),
+                        ch_length[ch])
+                except AssertionError:  # changed length or sequence structure
+                    for c in self.channel_groups[ch]:
+                        channels_to_program.append(c)
+            # update the length cache
+            for ch in channels_to_program:
+                sequence_cache['length'][ch] = ch_length.get(ch, {})
+            # Update the cache for channels that are on an AWG marked for
+            # complete re-programming (these channels might have been skipped
+            # above).
+            for ch in self.channels:
+                if self.get(f'{ch}_awg') in awgs_to_program:
+                    sequence_cache['settings'][ch] = settings.get(ch, {})
+                    sequence_cache['hashes'][ch] = channel_hashes.get(
+                        ch, {})
+                    sequence_cache['metadata'][ch] = metadata.get(ch, {})
+                    sequence_cache['length'][ch] = ch_length.get(ch, {})
+            log.debug(f'awgs_to_program = {repr(awgs_to_program)}\n'
+                      f'awgs_with_channels_to_upload = '
+                      f'{repr(awgs_with_channels_to_upload)}\n'
+                      f'channels_to_upload = {repr(channels_to_upload)}\n'
+                      f'channels_to_program = {repr(channels_to_program)}'
+                      )
```

Comment:
This part of the code is too complicated for me to review functionality-wise, which should be still ok as we have tested it a lot. Formatting- / structure-wise it looks ok.